### ISSUE TITLE
Streaming Audio Input.

### ISSUE THREAD

**Issue by Praj-17** [TMElyralab/MuseTalk#239]
Hi, I am generating Audio streams, and I want to pass the generated audio stream directly to the MuseTalk real-time inference script. and get realtime streams generated. Is it possible? If Yes, How?

**Comment by Viktoo** [comment]
I need this as well. But I don't think it's possible

For example, when inference begins, the code extracts features from audio found in a local audio_path. This implementation assumes the full audio file is ready at the start.
https://github.com/TMElyralab/MuseTalk/blob/058f7ddc7fa3eb59a7e05e71e1ae7712cdd1ae2b/musetalk/whisper/audio2feature.py#L97

**Comment by chuong-phung** [comment]
Should we split the audio into smaller chunk?

**Comment by rizwanishaq** [comment]
Anybody has done in the realtime, I am trying the whisper encoder like that I don't know if this is the correct way of doing 

What I am doing, I am getting 200msec buffer, and calculating the audio_features, is anybody has done something like that for realtime? 

```
# whisper_embedder.py

import torch
import numpy as np
from transformers import WhisperModel, WhisperFeatureExtractor
from positional_encoding import PositionalEncoding


class WhisperEmbedder:
    """
    Extracts fixed-shape Whisper encoder embeddings from raw audio segments.

    Each segment must be 3200 float32 samples (200ms @ 16kHz).
    Outputs a tensor of shape [B, T, L, D] = [batch_size, 10, 5, 384],
    where:
        - T = target time frames
        - L = target encoder layers (from bottom)
        - D = embedding dimension (384 for tiny model)
    """

    def __init__(
        self,
        model_name: str = "openai/whisper-tiny",
        cache_dir: str = "/models/audio_encoder/1/whisper",
        target_frames: int = 10,
        target_layers: int = 5,
        device = "cuda",
        dtype = torch.float16,
        logger=None
    ):
        """
        Initializes the Whisper encoder and feature extractor.

        Args:
            model_name (str): Hugging Face model name or local path.
            cache_dir (str): Directory for caching models locally.
            target_frames (int): Number of time frames to keep in output (T).
            target_layers (int): Number of encoder layers to extract (L).
        """
        self.device = device
        self.dtype = dtype
        self.target_frames = target_frames
        self.target_layers = target_layers

        # Load pretrained Whisper components
        self.feature_extractor = WhisperFeatureExtractor.from_pretrained(model_name, cache_dir=cache_dir)
        self.model = WhisperModel.from_pretrained(model_name, cache_dir=cache_dir)
        self.model.to(self.device, dtype=self.dtype).eval()


        # Compile the encoder forward pass
        self.model.encoder = torch.compile(self.model.encoder)

        # Pre-store positional encoding on GPU in correct dtype
        self.pe = PositionalEncoding(
            d_model=384,
            max_len=target_frames * target_layers,
            device=self.device,
            dtype=self.dtype
        )

        self.logger = logger


        # Run warmup so the first real inference is fast
        self.warmup()

    def warmup(self):
        """Run a dummy pass to trigger torch.compile graph building."""
        with torch.inference_mode():
            dummy_audio = np.zeros((1, 3200), dtype=np.float32)
            _ = self.extract_embeddings(dummy_audio)
    
    

    @torch.inference_mode()
    def extract_embeddings(self, audio_batch: np.ndarray) -> np.ndarray:
        """
        Extracts Whisper encoder embeddings from a batch of audio segments.

        Args:
            audio_batch (np.ndarray): Input array of shape [B, 3200] with float32 PCM samples.

        Returns:
            np.ndarray: Embeddings of shape [B, T, L, 384] as float32.
        """

        
        # Extract input features using the Whisper feature extractor
        features = self.feature_extractor(
            audio_batch,
            sampling_rate=16000,
            return_tensors="pt"
        ).input_features.to(self.device, dtype=self.dtype)
```
        
        
        # Run encoder and get all hidden states: [B, T, L=number of layers, D]
        encoder_outputs = self.model.encoder(features, output_hidden_states=True)
        hidden_states = encoder_outputs.hidden_states

        # Stack hidden states into a 4D tensor: [B, T, L, D]
        embeddings = torch.stack(hidden_states, dim=2)  # dim=2 -> layer axis

        # Keep only bottom 'target_layers' (e.g. first 5)
        embeddings = embeddings[:, :self.target_frames, :self.target_layers, :]  # [B, T, L, D]

        # Flatten T and L dimensions: [B, T*L, D] → [B, 50, 384]
        embeddings = embeddings.reshape(embeddings.size(0), -1, 384)
        embeddings = self.pe(embeddings)


        # Move to CPU and convert to NumPy
        return embeddings.cpu().numpy().astype(np.float32)
